---
title: "Differential expression - chemobrain mouse mRNA"
author: "Kimberly Olney, PhD"
date: "July 21st 2025"
output:
  pdf_document: default
---

Compare MTX versus saline:
MTX 150
MTX 300
MTX 450

# Set up working enivornment
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = ".")
getwd()
```

```{r libraries, message=FALSE, warning=FALSE}
source("/tgen_labs/jfryer/kolney/chemobrain/scripts/file_paths_and_colours.R")
```

# Create a counts matrix
This is the raw counts 
```{r raw_counts}
# path to counts files
count_files <-
  file.path(paste0(
    "/tgen_labs/jfryer/cores/tgen/20250714_LH00295_0255_A23377MLT3/BAM/grandcanyon_results/CB_all/", metadata$SampleID,
    ".genes.sf"
  ))
# add sample counts name to the counts files
names(count_files) <- paste0(metadata$Study_Specimen_ID)
# Inspect 
table(metadata$SampleID) # Some samples have technical replicates. 

# Create the counts matrix 
count_data <- data.frame(fread(count_files[1]))[c(1,5)] # RAW COUNTS
# Loop and read the 5th column
for(i in 2:length(count_files)) {
        count_data <- cbind(count_data, data.frame(fread(count_files[i]))[5])
}
# set gene_id as the row name
row.names(count_data) <- count_data$Name
count_data$Name <- NULL
```

# Format counts table
```{r counts_table}
# set column names to sample counts file name
colnames(count_data) <- metadata$Study_Specimen_ID
genes <- read.delim("/tgen_labs/jfryer/kolney/chemobrain/ensembl_mouse_genes.txt")
```

# Check inputs
```{r check_inputs}
count_gene_id <- rownames(count_data)
count_gene_id_df <- as.data.frame(count_gene_id)
colnames(count_gene_id_df) <- "GENEID"
genes2 <- merge(count_gene_id_df, genes, by = "GENEID", all = TRUE, all.x = TRUE, all.y = FALSE)
genes2_sorted <- genes2[match(count_gene_id_df$GENEID, genes2$GENEID), ]

all.equal(rownames(count_data), genes2_sorted$GENEID) # gene ids match the count rows
all.equal(colnames(count_data), (metadata$Study_Specimen_ID)) # counts columns match the sample IDs
```

# Create DGE object
```{r DGE_object}
# create object
count_data <- na.omit(count_data)
dge <- DGEList(counts = count_data,
               samples = metadata,
               genes = genes2_sorted)

# Inspect
table(dge$samples$group)
table(dge$samples$Study_Specimen_ID)
```

# Set factors and colors 
```{r factor_and_color_set}
dge$samples$group <- factor(dge$samples$group, levels = c("Saline", "MTX_150", "MTX_300", "MTX_450"))

condition_colors <- c("grey", "lightblue", "blue", "blue4")[dge$samples$group]
```

# Raw CPM
```{r raw_cpm}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
```

# Raw MDS
```{r raw_MDS}
# Condition
par(bg = 'white')
plotMDS(
  lcpm,
  top = 100,
  labels = dge$samples$Study_Specimen_ID,
  cex = 1,
  dim.plot = c(1, 2),
  plot = TRUE,
  col = condition_colors
)
title(expression('Top 500 Genes - Raw (Log'[2] ~ 'CPM)'))
saveToPDF("../results/MDS/MDS_raw_condition_salmon.pdf", width = 6, height = 6)

```

# JSD heatmap
This portion won't display in the R Markdown pdf; the margins are too large.
The pdf and png file can only be saved one at a time.
```{r JSD, warning = FALSE, eval=FALSE}
# save
path <- ("../results/JSD/JSD_raw")
pdf(paste0(path,".pdf"), width = 10, height = 8, pointsize = 8)

# set heatmap colors and names
colors <- c("blue","skyblue","white") # set heatmap color scale
colors <- colorRampPalette(colors)(100) # make it a gradient
sample_group_color <- c("grey", "lightblue", "blue", "blue4")[dge$samples$group]
names <- paste(dge$samples$Study_Specimen_ID)

# find JSD
data <- JSD(t(edgeR::cpm(dge$counts)), est.prob = "empirical")
colnames(data) <- names
rownames(data) <- names
round.data <- round(data, digits = 2) # round 2 decimal places

# plot heatmap
heatmap <- heatmap.2(
  round.data,
  trace = "none",
  colCol = sample_group_color,
  colRow = sample_group_color,
  symm = TRUE,
  col = colors,
  cellnote = round.data,
  notecex = 1,
  dendrogram = "none",
  notecol = "black",
  key.title = "Color Key",
  srtCol = 65,
  margins = c(12,12),
  keysize = 0.2)

# clean up
rm(data, round.data, colors, path, names)
```

# Remove mitochondrial genes
```{r MT_genes}
dim(dge)
dge$samples$lib.size
removeMT <- dge$genes$Chr != "MT"  # true when NOT MT
dge <- dge[removeMT,,keep.lib.sizes = TRUE]
dim(dge)
```

# Keep only protein coding genes 
```{r MT_genes}
dim(dge)
removeMT <- dge$genes$gene_biotype == "protein_coding"  # true when NOT MT
dge <- dge[removeMT,,keep.lib.sizes = TRUE]
dim(dge)
```

# Library sizes
```{r library}
# before filtering
L <- mean(dge$samples$lib.size) * 1e-6
M <- median(dge$samples$lib.size) * 1e-6
c(L, M)
```

# Filtering 
The filterByExpr() function in the edgeR package determines which genes have a great enough count value to keep.  
We will filter by group.  
This means at least 3 samples (3 is the smallest group sample size) must express a minimum count of 1 CPM. 
```{r filter_counts}
# first filter by expression
dim(dge)
keep.expr <-
  filterByExpr(
    dge,
    group = dge$samples$group, # by disease groups
    min.count = M, # min count of 1 CPM 
  )
dge.filtered <- dge[keep.expr, ,keep.lib.sizes = TRUE]
dim(dge.filtered)
table(dge.filtered$genes$gene_biotype)
dge.filtered$samples$lib.size
```

# TMM
For estimating relative RNA production levels from RNA-seq data. 
The TMM method estimates scale factors between samples that can be incorporated 
into currently used statistical methods for DE analysis.
```{r TMM}
# Now, method of trimmed mean of M-values (TMM)
# Calculate scaling factors to convert raw library sizes into effective library sizes.
dge.filtered$counts <- na.omit(dge.filtered$counts)
dge.filtered.norm <- calcNormFactors(dge.filtered$counts, method = "TMM")
```

# Gene_id to gene_name
Some of the gene_names are duplicated. Our counts matrix has rownames by gene_id which are not duplicated. Since we have now filtered and normalized we can check to see if 
duplicated gene_names still exist.  If none do, rename rows to gene_name.  If there are duplicates, handle the error and then rename columns.
```{r gene_name}
# check for duplication
table(duplicated(dge.filtered.norm$genes$gene_name))

# replace NA with gene_id 
dge.filtered.norm$genes <- dge.filtered.norm$genes %>% mutate(gene_name = coalesce(gene_name,gene_id))

# check for duplication
table(duplicated(dge.filtered.norm$genes$gene_name))
```
# Clean up
```{r cleanup}
rm(count_data, lcpm, keep.expr, normfactors, removeMT, condition_colors, L, M, typeOfCount)
```

# Filtered CPM
```{r filtered_lcpm}
lcpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE)
```

# Filtered MDS
```{r filtered_MDS}
# Condition
par(bg = 'white')
plotMDS(
  lcpm,
  top = 500,
  labels = dge.filtered.norm$samples$group,
  cex = 1,
  dim.plot = c(1, 2),
  plot = TRUE,
  col = condition_colors
)
title(expression('Top 500 Genes - Filtered (Log'[2] ~ 'CPM)'))
saveToPDF("../results/MDS/MDS_filtered_condition.pdf", width = 6, height = 6)
```

# Density plot
Density plots of log - intensity distribution of each library can be superposed on a single graph for a better comparison between libraries and for identification of libraries with weird distribution. 
```{r density_plots}
# set graphical parameter
par(mfrow = c(1,3))

# Normalize data for library size and expression intensity
log2cpm.tech <- edgeR::cpm(dge, log = TRUE)
log2cpm.filtered <- edgeR::cpm(dge.filtered, log = TRUE)
log2cpm.norm <- edgeR::cpm(dge.filtered.norm, log = TRUE)

# set colors
colors <- condition_colors
nsamples <- ncol(dge)

# First, plot the first column of the log2cpm.tech density
plot(density(log2cpm.tech[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "A. Raw", xlab = expression('Log'[2]~CPM))
# For each sample plot the lcpm density
for (i in 1:nsamples){
  den <- density(log2cpm.tech[,i]) #subset each column
  lines(den$x, den$y, col = colors[i], lwd = 2) 
}

# Second, plot log2cpm.filtered
plot(density(log2cpm.filtered[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "B. Filtered", xlab = expression('Log'[2]~CPM))
for (i in 2:nsamples) {
  den <- density(log2cpm.filtered[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# Third, plot log2cpm.norm
plot(density(log2cpm.norm[,1]), col = colors[1], lwd = 2, ylim = c(0,0.5), 
     las = 2, main = "C. TMM", xlab = expression('Log'[2]~CPM))
for (i in 2:nsamples) {
  den <- density(log2cpm.norm[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# save
path <- ("../results/library/gene_density")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 5)
```

# Boxplots
```{r boxplots}
# set parameters
par(mfrow = c(1, 3))

# First look at dge.tech
boxplot(
  log2cpm.tech,
  main = "A. Raw",
  xlab = "",
  ylab = expression('Counts per gene (Log'[2] ~ 'CPM)'),
  axes = FALSE,
  col = colors
)
axis(2) # 2 = left
axis(
  1,
  # 1 = below
  at = 1:nsamples,
  # points at which tick-marks should be drawn
  labels = colnames(log2cpm.tech),
  las = 2,
  cex.axis = 0.8 # size of axis
)

# Second, look at dge.filtered
boxplot(
  log2cpm.filtered,
  main = "B. Filtered",
  xlab = "",
  ylab = expression('Counts per gene (Log'[2] ~ 'CPM)'),
  axes = FALSE,
  col = colors
)
axis(2)
axis(
  1,
  at = 1:nsamples,
  labels = colnames(log2cpm.filtered),
  las = 2,
  cex.axis = 0.8
)

# Third, look at dge.norm
boxplot(
  log2cpm.norm,
  main = "C. Normalized",
  xlab = "",
  ylab = expression('Counts per gene (Log'[2] ~ 'CPM)'),
  axes = FALSE,
  col = colors
)
axis(2)
axis(
  1,
  at = 1:nsamples,
  labels = colnames(log2cpm.norm),
  las = 2,
  cex.axis = 0.8
)
path <- ("../results/library/lcpm_boxplots")
saveToPDF(paste0(path, ".pdf"), width = 11, height = 5)

# clean up
rm(log2cpm.tech, log2cpm.norm, log2cpm.filtered, den, colors, nsamples, path, i)
```

# save filtered dge
```{r save_dge_filtered_norm}
saveRDS(dge.filtered.norm, file = "../rObjects/dge.filtered.norm.4vs4.rds")
rm(dge, dge.tech, dge.filtered, lcpm, tx2gene, condition_colors)
```

Voom transform counts to use for variancePartition  
```{r voom, warning=FALSE}
formula <- (~ 0 + group)
counts <- as.data.frame(dge.filtered$counts)
info <- as.data.frame(dge.filtered$samples)
voom_with_weights <-
  variancePartition::voomWithDreamWeights(
    counts = counts,
    formula = formula,
    data = info,
    BPPARAM = BiocParallel::SnowParam(8),
    plot = TRUE
  )
path <- paste0("../results/voom/voom_raw")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
voomCounts <- voom_with_weights$E
rm(voom_with_weights, formula)
```

# Fit variance 
variancePartition quantifies and interprets multiple sources of biological and technical variation in gene expression experiments. The package a linear mixed model to quantify variation in gene expression attributable to individual, tissue, time point, or technical variables.
```{r varpart}
form_varPart <- ~ (1|group)
# fit model and extract variance percents
rownames(info) <- info$pig_id
varPart <- fitExtractVarPartModel(voomCounts, form_varPart, info)
```

# Plot VarPart
```{r plot_varPart}
plotVarPart(sortCols(varPart), label.angle = 80)
saveToPDF("../results/variance/variance_partition_violins.pdf", width = 8.5, height = 5)

varPart$gene_id <- rownames(varPart)
# merge with gene information to get gene names for gene_id
variance_explained <- merge(varPart, genes, by = "gene_id")

write.table(variance_explained, "../results/variance/variance_explained.tsv",sep = "\t", quote = FALSE, row.names = FALSE)
```

# Save varPart
```{r save_varPart}
saveRDS(varPart,   "../rObjects/varpart.rds")
rm(varPart, variance_explained)
```


# CCA 
Canonical Correlation Analysis
```{r}
form <- ~ condition_dose  + duration_min + age_days + end_heart_rate + end_temp
# Compute Canonical Correlation Analysis (CCA) # between all pairs of variables
# returns absolute correlation value
C = canCorPairs(form, info)
# Plot correlation matrix
plotCorrMatrix( C )
saveToPDF("../results/variance/CCA_with_project.pdf", width = 8, height = 8)

rm(form, C)
```

# Design matrix
```{r design_matrix}
design <- model.matrix( ~ 0 + info$group)
colnames(design) <- c("Saline", "MTX_150", "MTX_300", "MTX_450")
# Inspect
design
```

# Voom
When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. 
The voom method estimates the mean-variance relationship of the log-counts.
Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
```{r voom_covariates}
form <- ~ 0 + group + Sex
voom_cov <-
  variancePartition::voomWithDreamWeights(
    counts = counts,
    formula = form,
    data = info,
    BPPARAM = BiocParallel::SnowParam(8),
    plot = TRUE
  )
voomCounts <- voom_cov$E
```

# Contrast plot
```{r contrasts}
# fits linear model for each gene given a series of arrays
fit <- lmFit(voom_cov, design)
coef.fit <- fit$coefficients

contrasts <- makeContrasts(
  MTX_150_vs_Saline = MTX_150 - Saline,
  MTX_300_vs_Saline = MTX_300 - Saline,
  MTX_450_vs_Saline = MTX_450 - Saline,
  levels = colnames(design))
head(contrasts)

# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)

# Compute differential expression based on the empirical Bayes moderation of the standard errors towards a common value.
# The logCPM values can then be used in any standard limma pipeline, using the trend=TRUE
veBayesFit <- eBayes(vfit, trend = TRUE, robust=TRUE)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")

rm(design, fit, coef.fit, contrasts, vfit)
```

# DEGs summary
```{r DGE_summary}
qval <- 0.05
lfc.cutoff <- 1

sumTable <- 
  summary(decideTests(
    vfit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = qval,
    lfc = lfc.cutoff  # numeric, minimum absolute log2-fold change required
  ))

print(paste0(" q < ", qval, " & absolute log2-fold change > ", lfc.cutoff))
sumTable
write.table(sumTable, paste0("../results/DEGs/DEGs_summary_q0.05_log2FC1.txt"), quote = FALSE, sep = "\t")
genes_relevant <- dge.filtered.norm$genes

rm(sumTable)
```

# DGE Check 
```{r DGE_check, eval=FALSE}
DGE_test <- topTable(
  veBayesFit, 
  coef = "Ecoli_vs_Saline",  
  n = Inf, 
  p.value = 1,
  lfc = 0, 
  sort.by = "P", 
  genelist = dge.filtered.norm$genes, 
  confint = TRUE # column of confidence interval 
    )
```

# Output DEG tables
```{r output_DEG_tables}
coef <- 1

for (i in allComparisons) {
  vTopTableAll <- topTable(
    veBayesFit, 
    coef = coef,  
    n = Inf, 
    p.value = 1,
    lfc = 0, 
    sort.by = "P", 
    genelist = genes_relevant, 
    confint = TRUE # column of confidence interval 
    )
    saveRDS(vTopTableAll, file = 
            paste0("../rObjects/gene_tables/", i,"_gene_table.rds"))
  path <- paste0("../results/DEGs/", i, "_q1.00.txt", sep = "") 
  write.table(vTopTableAll, path, sep = "\t", row.names = FALSE, quote = FALSE)
  
  # p < 0.05, log2fc > 1
  vTopTableSig <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = qval,
      lfc = lfc.cutoff,
      genelist = genes_relevant, 
      confint = TRUE)
  path <- paste0("../results/DEGs/", i, "_q0.05_lfc1.txt", sep = "") 
  write.table(vTopTableSig, path, sep = "\t", row.names = FALSE, quote = FALSE)
  # increment 
  coef <- coef + 1
}
remove(coef, vTopTableAll, vTopTableSig)
```

# PCA
Principal component analysis, or PCA, is a dimensionality reduction method that is often used to reduce the dimensionality of large data sets, by transforming a large set of variables into a smaller one that still contains most of the information in the large set.
```{r PCA}
# Setting the N of genes to use
ntop = length(dge.filtered.norm$genes$gene_id)
# Sorting by the coefficient of variance
means <- rowMeans(voomCounts)
Pvars <- rowVars(voomCounts, useNames = TRUE)
cv2 <- Pvars / means ^ 2
select <- order(cv2, decreasing = TRUE)[seq_len(min(ntop, length(cv2)))]
highly_variable_exp <- ((voomCounts)[select,])
dim(highly_variable_exp)

# Running PCA
pca_exp <- prcomp(t(highly_variable_exp), scale = F, center = T)
# scale a logical value indicating whether the variables should be scaled to have unit variance before the analysis takes place.
# a logical value indicating whether the variables should be shifted to be zero centered.

# Dataframe with the first 8 PCs
dim1_10 <- data.frame(pca_exp$x[, 1:8])
# Adding metadata
dim1_10$sample <- rownames(dim1_10)
pcaWithMetadata <- merge(dim1_10, info, by = "sample", all = TRUE)
pcaWithMetadata$group <- pcaWithMetadata$condition

# Plotting
ggplot(data = pcaWithMetadata, aes(x = PC1, y = PC2, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw() 
saveToPDF("../results/PCA/PCA_dim1&2.pdf", width = 6, height = 6)

ggplot(data = pcaWithMetadata, aes(x = PC2, y = PC3, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw() 
saveToPDF("../results/PCA/PCA_dim2&3.pdf", width = 6, height = 6)

write.table(pcaWithMetadata, "../results/PCA/PCA_metadata.txt", quote = FALSE, sep = "\t")
```

# CCA with PC1-10 & variables in model 
```{r CCA_PCA}
form_PCA <- ~ condition + duration_min + age_days + end_heart_rate + end_temp + 
  PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 
C = canCorPairs(form_PCA, pcaWithMetadata)
plotCorrMatrix( C )
saveToPDF("../results/variance/CCA_with_PCA.pdf", width = 8, height = 8)

rm(dim1_10, highly_variable_exp, pcaWithMetadata, pca_exp, voomCounts, cv2, means, ntop, path, select, Pvars, C, form_PCA)
```

# Volcano plots
```{r volcano, warning=FALSE}
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(paste0("../results/DEGs/", i,"_q1.00.txt"))
    i <- gsub("_", " ", i)
    if (!any(group1_vs_group2$adj.P.Val < 0.05, na.rm = TRUE)) {
    # If there are no significant DEGs, create a volcano plot with all points in gray
    p <- ggplot(data = group1_vs_group2, aes(x = logFC, y = -log10(P.Value))) +
      geom_point(alpha = 0.8, size = 2, color = "gray") +
      theme_bw() +
      theme(legend.position = "none",
            axis.title.x = element_text(size = 10),
            axis.text.x = element_text(size = 10),
            axis.title.y = element_text(size = 10),
            axis.text.y = element_text(size = 10)) +
      geom_vline(xintercept = negFC,
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 colour = "#000000",
                 linetype = "dashed") +
      labs(x = expression(log[2](FC)),
           y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")) +
      ggtitle(paste0(i, "\nNo significant DEGs found"))
    i <- gsub(" ", "_", i)
    pdf(paste0("../results/volcano/", i, ".pdf"), height = 6, width = 6)
    print(p)
    dev.off()
  } else {
  color_values <- vector()
  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$adj.P.Val[row] < 0.05) {
      if (group1_vs_group2$logFC [row] > 1) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$logFC[row] < -1) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  # plot only if there are DEGs with p_val_adj < 0.05
  num <- subset(data, (adj.P.Val < 0.05 & logFC < -1)  | (adj.P.Val < 0.05 & logFC > 1 ))
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, logFC > 1)
    upFold <- upFold[!(upFold$gene_name %in% up10$gene_name),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, logFC < -1)
    downFold <- downFold[!(downFold$gene_name %in% down10$gene_name),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$P.Value[data$adj.P.Val < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-1)
    posFC <- c(1) 
    p <-
      ggplot(data = data,
             aes(
               x = logFC,
               y = -log10(P.Value),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = 1) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i, "\nq < 0.05 & |log2FC| > 1")) +
      geom_text_repel(
        data = up10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = upFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = down10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = downFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      )
    p
    i <- gsub(" ", "_", i)
    # save
    path <-paste0("../results/volcano/",i)
    pdf(paste0(path, ".pdf"),
        height = 6,
        width = 6)
    print(p)
    dev.off()
    
    print(paste("i =", i))
  }
  } 
}
```

```{r cleanup}
# clean up
remove(up, up10, upFold, group1_vs_group2, downFold, down10, data, p, my_colors)
```

# Make excel table
```{r excel}
# first read in the files
for (i in allComparisons) {
  filepath <- paste0("../results/DEGs/", i,"_q1.00.txt")
  assign(paste0(i),
         read.delim(filepath, header = TRUE, sep = "\t",stringsAsFactors = FALSE))
}

allComparisons
list_of_datasets <- list(
  "Ecoli_vs_Saline" = Ecoli_vs_Saline)
write.xlsx(list_of_datasets, file = "../results/DEGs/DEGs.q1.00.xlsx")

library("readxl")
DEGs <- read_excel("../results/DEGs/DEGs.q1.00.xlsx")
SIG_DEGs <- subset(DEGs, (adj.P.Val < 0.05 & logFC > 1) | (adj.P.Val < 0.05 & logFC < -1))
write.table(SIG_DEGs, file = "../results/DEGs/Ecoli_vs_Saline_q0.05_lfc1.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

```{r}
rownames(veBayesFit$p.value) <- dge.filtered.norm$genes$gene_name
rownames(veBayesFit$coefficients) <- dge.filtered.norm$genes$gene_name
rownames(veBayesFit$stdev.unscaled) <- dge.filtered.norm$genes$gene_name
#rownames(veBayesFit$cov.coefficients) <- dge.filtered.norm$genes$gene_name
rownames(veBayesFit$t) <- dge.filtered.norm$genes$gene_name
rownames(veBayesFit$lods) <- dge.filtered.norm$genes$gene_name


glimmaVolcano(veBayesFit, dge = dge.filtered.norm, groups = dge.filtered.norm$samples$group)


#sessionInfo()
```
